src/Dialect/Mlir/DialectBuilder.cpp

```diff
@@ -69,56 +70,104 @@ namespace onnx_mlir {
   return elementType;
 }
 
-/* static */ bool MathBuilder::isIntegerWithVector(Type elementOrVectorType) {
-  Type elementType = elementTypeWithVector(elementOrVectorType);
+/* static */ bool MathBuilder::isScalarOrVectorInteger(
+    Type elementOrVectorType) {
+  Type elementType = elementTypeOfScalarOrVector(elementOrVectorType);
   return mlir::isa<IntegerType>(elementType) ||
          mlir::isa<IndexType>(elementType);
 }
 
-/* static */ bool MathBuilder::isUnsignedIntegerWithVector(
+/* static */ bool MathBuilder::isScalarOrVectorUnsignedInteger(
     Type elementOrVectorType) {
-  Type elementType = elementTypeWithVector(elementOrVectorType);
+  Type elementType = elementTypeOfScalarOrVector(elementOrVectorType);
   return elementType.isUnsignedInteger();
 }
 
-/* static */ bool MathBuilder::isFloatWithVector(Type elementOrVectorType) {
-  Type elementType = elementTypeWithVector(elementOrVectorType);
+/* static */ bool MathBuilder::isScalarOrVectorFloat(Type elementOrVectorType) {
+  Type elementType = elementTypeOfScalarOrVector(elementOrVectorType);
   return mlir::isa<FloatType>(elementType);
 }
 
+bool MathBuilder::splatToMatch(Value &first, Value &second) const {
+  bool firstIsVector = isVector(first.getType());
+  bool secondIsVector = isVector(second.getType());
+  // Both scalar or vector, nothing to do.
+  if (firstIsVector == secondIsVector) {
+    // If both vectors, ensure its the same vector length.
+    if (firstIsVector) {
+      assert(
+          VectorBuilder::compatibleTypes(first.getType(), second.getType()) &&
+          "expected compatible types");
+    }
+    // No changes in types.
+    return false;
+  }
+  if (firstIsVector) {
+    // Splat second.
+    assert(!secondIsVector && "bad assumption");
+    VectorType firstVectorType = mlir::cast<VectorType>(first.getType());
+    VectorType secondVectorType =
+        VectorType::get(firstVectorType.getShape(), second.getType());
+    second = b().create<vector::SplatOp>(loc(), secondVectorType, second);
+  } else {
+    // Splat first.
+    assert(secondIsVector && "bad assumption");
+    VectorType secondVectorType = mlir::cast<VectorType>(second.getType());
+    VectorType firstVectorType =
+        VectorType::get(secondVectorType.getShape(), first.getType());
+    first = b().create<vector::SplatOp>(loc(), firstVectorType, first);
+  }
```

Comment:
The current code is good as it is. FYI, I just tried to rewrite this to make it more natural to read, e.g.: `if first is not a vector, splat it to a vector, if second is not a vector, splat it to a vector`, instead of `if first is a vector, splat second to a vector, ...`.

```C
VectorType firstVectorType = mlir::dyn_cast<VectorType>(first.getType());
VectorType secondVectorType = mlir::dyn_cast<VectorType>(second.getType());

if (!firstVectorType && secondVectorType) {
  firstVectorType = VectorType::get(secondVectorType.getShape(), first.getType());
  first = b().create<vector::SplatOp>(loc(), firstVectorType, first);
}

if (firstVectorType && !secondVectorType) {
 secondVectorType = VectorType::get(firstVectorType.getShape(), second.getType());
 second = b().create<vector::SplatOp>(loc(), secondVectorType, second);
}
```

Reply:
I like it, thanks for rewriting my code! Better than any chatbots could ever be :-)

Reply:
:) but I guess there was an error in my code, I should not update `firstVectorType` but use a local variable for it.

```C 
VectorType firstVectorType = mlir::dyn_cast<VectorType>(first.getType());
VectorType secondVectorType = mlir::dyn_cast<VectorType>(second.getType());

if (!firstVectorType && secondVectorType) {
  VectorType ty = VectorType::get(secondVectorType.getShape(), first.getType());
  first = b().create<vector::SplatOp>(loc(), ty, first);
}

if (firstVectorType && !secondVectorType) {
 VectorType ty = VectorType::get(firstVectorType.getShape(), second.getType());
 second = b().create<vector::SplatOp>(loc(), ty, second);
}
```

Reply:
I did not notice it, but in fact I changed the code a bit so that it actually does not matter. Here is the latest code
```C++
bool MathBuilder::splatToMatch(Value &first, Value &second) const {
  Type firstType = first.getType();
  Type secondType = second.getType();
  VectorType firstVectorType = mlir::dyn_cast<VectorType>(firstType);
  VectorType secondVectorType = mlir::dyn_cast<VectorType>(secondType);
  VectorBuilder createVec(*this);
  // Splat first if needed.
  if (!firstVectorType && secondVectorType) {
    firstVectorType = VectorType::get(secondVectorType.getShape(), firstType);
    first = createVec.splat(firstVectorType, first);
    return true;
  }
  // Splat second if needed.
  if (firstVectorType && !secondVectorType) {
    secondVectorType = VectorType::get(firstVectorType.getShape(), secondType);
    second = createVec.splat(secondVectorType, second);
    return true;
  }
  // Otherwise check compatibility.
  assert(createVec.compatibleTypes(firstType, secondType) &&
         "expected compatible types");
  return false;
}
```